shared/typeinference/codeql/typeinference/internal/TypeInference.qll

```diff
@@ -0,0 +1,776 @@
+            or
+            exists(TypePath prefix, TypePath suffix, TypeParameter tp |
+              baseTypeMentionHasTypeParameterAt(sub, baseMention, prefix, tp) and
+              t = inferTypeAt(a, apos, tp, suffix) and
+              path = prefix.append(suffix)
+            )
+          )
+        }
+      }
+
+      pragma[nomagic]
+      private predicate accessBaseType(
+        Access a, AccessPosition apos, Declaration target, Type base, TypePath path, Type t
+      ) {
+        exists(TypeMention tm |
+          target = a.getTarget() and
+          AccessBaseType::hasBaseTypeMention(a, apos, tm, path, t) and
+          base = resolveTypeMentionRoot(tm)
+        )
+      }
+
+      pragma[nomagic]
+      private predicate declarationBaseType(
+        Declaration decl, DeclarationPosition dpos, Type base, TypePath path, Type t
+      ) {
+        t = decl.getDeclaredType(dpos, path) and
+        path.isCons(base.getATypeParameter(), _)
+      }
+
+      /**
+       * Holds if the (transitive) base type `t` at `path` of `a` matches the type
+       * parameter `tp`, which is used in the declared types of `target`.
+       *
+       * For example, in
+       *
+       * ```csharp
+       * class C<T1> { }
+       *
+       * class Base<T2> {
+       * //         ^^ `tp`
+       *     public C<T2> Method() { ... }
+       * //               ^^^^^^ `target`
+       * }
+       *
+       * class Mid<T3> : Base<C<T3>> { }
+       *
+       * class Sub<T4> : Mid<C<T4>> { }
+       *
+       *    new Sub<int>().Method();
+       * // ^^^^^^^^^^^^^^^^^^^^^^^ `a`
+       * ```
+       *
+       * we have that type parameter `T2` of `Base` is matched as follows:
+       *
+       * `path`    | `t`
+       * --------- | -------
+       * `"0"`     | ``C`1``
+       * `"0.0"`   | ``C`1``
+       * `"0.0.1"` | `int`
+       */
+      pragma[nomagic]
+      private predicate baseTypeMatch(
+        Access a, Declaration target, TypePath path, Type t, TypeParameter tp
+      ) {
+        exists(AccessPosition apos, DeclarationPosition dpos, Type base, TypePath pathToTypeParam |
+          accessBaseType(a, apos, target, base, pathToTypeParam.append(path), t) and
+          declarationBaseType(target, dpos, base, pathToTypeParam, tp) and
+          not exists(getTypeArgument(a, target, tp, _)) and
+          accessDeclarationPositionMatch(apos, dpos)
+        )
+      }
+
+      pragma[nomagic]
+      private predicate explicitTypeMatch(
+        Access a, Declaration target, TypePath path, Type t, TypeParameter tp
+      ) {
+        target = a.getTarget() and
+        t = getTypeArgument(a, target, tp, path)
+      }
+
+      pragma[nomagic]
+      private predicate implicitTypeMatch(
+        Access a, Declaration target, TypePath path, Type t, TypeParameter tp
+      ) {
+        directTypeMatch(a, target, path, t, tp)
+        or
+        baseTypeMatch(a, target, path, t, tp)
```

Comment:
So `adjustAccessType` is applied in `directTypeMatch`, but not in `baseTypeMatch`? Maybe an example to clarify why we have this distinction would be good to add in the qldoc for `adjustAccessType`.

Reply:
Correct, only used in `directTypeMatch`. I think we would have to add it in `accessBaseType` as well.